### PR TITLE
Reset Periodic Vesting Accounts for Kava-8

### PR DESCRIPTION
**Pull request by nddeluca**
This PR removes all **vested** periods from periodic vesting accounts at the time genesis of kava-8.

This is done to optimize gas, transaction, and storage costs due to a high number of already vested periods for some users.

If a periodic vesting account is fully vested (has zero vested periods), the account will be converted back into a base account with no vesting information.

If the original vesting amount left is less than the delegated vesting amount, the delegated vesting amount is set to the original vesting value, and the change in delegated vesting is added to delegated free.  The total delegated funds does not change.

The unit tests cover all the edge cases I could think of in the vesting period and delegation handling, and the migrate tests assert the following:
- The number of accounts after the migration is the same
- There are no vested periods left for periodic vesting accounts.
- There are no periodic vesting accounts with zero vesting periods
- Delegated vesting is not greater than original vesting for all periodic vesting accounts
- The total owned coins (base coins) for all accounts is not modified
- The spendable coins at genesis time, 30 days after, 90 days after, 180 days after and 365 days after are equal for the unmigrated and migrated accounts.

### CODE REVIEW

File: migrate/v0_15/account_test.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+			Length: 0, // 0 days
+			Amount: sdk.NewCoins(sdk.NewCoin("ukava", sdk.NewInt(1e6))),
+		},
+		vesting.Period{
+			Length: 15 * 24 * 60 * 60, // 15 days
+			Amount: sdk.NewCoins(sdk.NewCoin("ukava", sdk.NewInt(1e6))),
+		},
+	}
+
+	assert.Equal(t, sdk.NewCoins(sdk.NewCoin("ukava", sdk.NewInt(2e6))), vacc.OriginalVesting, "expected original vesting to be unchanged")
```
Comment (karzak) [inline migrate/v0_15/account_test.go:221]: "expected original vesting to be changed". correct? Looks like it goes from 3e6 to 2e6

File: migrate/v0_15/account_test.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	spendableBefore := vacc.SpendableCoins(newVestingStartTime)
+
+	ResetPeriodicVestingAccount(vacc, newVestingStartTime)
+
+	// new period length 15 days
+	expectedEndtime := newVestingStartTime.Add(15 * 24 * time.Hour).Unix()
+	// new period length changed, amount unchanged
+	expectedPeriods := vesting.Periods{
+		vesting.Period{
+			Length: 0, // 0 days
```
Comment (karzak) [inline migrate/v0_15/account_test.go:212]: Bit of a nit/extreme edge case, but why keep a vesting period with length 0?

File: migrate/v0_15/account_test.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+		},
+	}
+
+	vacc := createVestingAccount(balance, vestingStartTime, periods)
+	genesisTime := vestingStartTime.Add(30 * 24 * time.Hour)
+
+	MigrateAccount(vacc, genesisTime)
+
+	assert.Equal(t, genesisTime.Unix(), vacc.StartTime, "expected vesting start time to equal genesis time")
+	assert.Equal(t, 1, len(vacc.VestingPeriods), "expected only one vesting period left")
```
Comment (karzak) [inline migrate/v0_15/account_test.go:72]: probably worth specifying the exact expected periods

File: migrate/v0_15/migrate_test.go [hunk 4]
```diff
@@ -99,3 +103,89 @@ func TestIncentive(t *testing.T) {
...
+func TestAuth_AccountConversion(t *testing.T) {
+	bz, err := ioutil.ReadFile(filepath.Join("testdata", "kava-7-test-auth-state.json"))
+	require.NoError(t, err)
+
+	var genesisState auth.GenesisState
+	cdc := app.MakeCodec()
+	cdc.MustUnmarshalJSON(bz, &genesisState)
+
+	migratedGenesisState := Auth(genesisState, GenesisTime)
+	require.Equal(t, len(genesisState.Accounts), len(migratedGenesisState.Accounts), "expected the number of accounts after migration to be equal")
```
Comment (karzak) [inline migrate/v0_15/migrate_test.go:148]: I would check that the resulting genesis is valid:

```
err = auth.ValidateGenesis(migratedGenesisState)
require.NoError(t, err)
```

File: migrate/v0_15/migrate_test.go [hunk 5]
```diff
@@ -99,3 +103,89 @@ func TestIncentive(t *testing.T) {
...
+			require.Equal(t, oldTotalDelegated, newTotalDelegated, "expected total amount of tracked delegations to not change")
+
+			// delegated vesting must be less or equal to original vesting
+			require.True(t, vacc.DelegatedVesting.IsAllLTE(vacc.OriginalVesting), "expected delegated vesting to be less or equal to original vesting")
+
+			// vested coins must be nil for the new account
+			require.Equal(t, sdk.Coins(nil), vacc.GetVestedCoins(GenesisTime), "expected no vested coins at genesis time")
+
+			// vesting coins must not be nil
+			require.NotEqual(t, sdk.Coins(nil), vacc.GetVestingCoins(GenesisTime), "expected vesting coins to be greater than 0")
```
Comment (karzak) [inline migrate/v0_15/migrate_test.go:188]: could maybe add an assertion here than the number of vesting periods on the new vesting account is less than or equal to the number of vesting periods on the old vesting account.

File: migrate/v0_15/migrate_test.go [hunk 6]
```diff
@@ -99,3 +103,89 @@ func TestIncentive(t *testing.T) {
...
+		futureDate = GenesisTime.Add(90 * 24 * time.Hour)
+		require.Equal(t, oldAcc.SpendableCoins(futureDate), acc.SpendableCoins(futureDate), "expected spendable coins to not chain")
+		// check 180 days
+		futureDate = GenesisTime.Add(180 * 24 * time.Hour)
+		require.Equal(t, oldAcc.SpendableCoins(futureDate), acc.SpendableCoins(futureDate), "expected spendable coins to not chain")
+		// check 365 days
+		futureDate = GenesisTime.Add(365 * 24 * time.Hour)
+		require.Equal(t, oldAcc.SpendableCoins(futureDate), acc.SpendableCoins(futureDate), "expected spendable coins to not chain")
+
+		if vacc, ok := acc.(vesting.PeriodicVestingAccount); ok {
```
Comment (karzak) [inline migrate/v0_15/migrate_test.go:171]:
```
if vacc, ok := acc.(*vesting.PeriodicVestingAccount)
```

File: migrate/v0_15/migrate_test.go [hunk 7]
```diff
@@ -99,3 +103,89 @@ func TestIncentive(t *testing.T) {
...
+			// old account must be a periodic vesting account
+			oldVacc, ok := oldAcc.(vesting.PeriodicVestingAccount)
```
Comment (karzak) [inline migrate/v0_15/migrate_test.go:173]: oldVacc, ok := oldAcc.(*vesting.PeriodicVestingAccount)

File: migrate/v0_15/migrate_test.go [hunk 8]
```diff
@@ -99,3 +103,89 @@ func TestIncentive(t *testing.T) {
...
+	cdc := app.MakeCodec()
+	cdc.MustUnmarshalJSON(bz, &genesisState)
+
+	migratedGenesisState := Auth(genesisState, GenesisTime)
+
+	assert.Equal(t, genesisState.Params, migratedGenesisState.Params, "expected auth parameters to not change")
+}
+
+func TestAuth_AccountConversion(t *testing.T) {
+	bz, err := ioutil.ReadFile(filepath.Join("testdata", "kava-7-test-auth-state.json"))
```
Comment (karzak) [inline migrate/v0_15/migrate_test.go:140]: Mostly curiosity, but do you know at what block this was exported?

**Review by karzak — CHANGES_REQUESTED** [review]
Core logic looks good, mostly minor suggestions

File: migrate/v0_15/account_test.go (hunk 2 above)
Comment (nddeluca) [reply migrate/v0_15/account_test.go:219]: It's still vesting at the newStartTime.  The comparison `currentTime < startTime + periodLength`.  If we do not keep the period with zero length, then the spendable balance will increase and not be equal to the previous spendable balance at newStartTime.

File: migrate/v0_15/migrate_test.go (hunk 8 above)
Comment (nddeluca) [reply migrate/v0_15/migrate_test.go:140]: No, looks like it was a current height export on 2021-08-05.  I can update to a specific height

**Review by karzak — APPROVED** [review]
LGTM
